packages/shorebird_cli/lib/src/commands/release_new/release_pipeline.dart

```diff
@@ -0,0 +1,347 @@
+
+        logger
+          ..success('✅ Published Release ${release.version}!')
+          ..info(postReleaseInstructions);
+
+        printPatchInstructions(releaseVersion: releaseVersion);
+      },
+      values: {
+        shorebirdEnvRef.overrideWith(() => releaseFlutterShorebirdEnv),
+      },
+    );
+  }
+
+  /// Asserts that the command can be run.
+  Future<void> validatePreconditions();
+
+  /// Asserts that the combination arguments passed to the command are valid.
+  Future<void> validateArgs();
+
+  /// Determines which Flutter version to use for the release. This will be
+  /// either the version specified by the user or the version provided by
+  /// [shorebirdEnv]. A [BuildPipelineException] will be thrown if the version
+  /// specified by the user is not found/supported.
+  Future<String> resolveTargetFlutterVersion() async {
+    if (flutterVersionArg != null) {
+      final String? revision;
+      try {
+        revision = await shorebirdFlutter.getRevisionForVersion(
+          flutterVersionArg!,
+        );
+      } catch (error) {
+        throw BuildPipelineException(
+          message: '''
+Unable to determine revision for Flutter version: $flutterVersionArg.
+$error''',
+          exitCode: ExitCode.software,
+        );
+      }
+
+      if (revision == null) {
+        final openIssueLink = link(
+          uri: Uri.parse(
+            'https://github.com/shorebirdtech/shorebird/issues/new?assignees=&labels=feature&projects=&template=feature_request.md&title=feat%3A+',
+          ),
+          message: 'open an issue',
+        );
+        throw BuildPipelineException(
+          message: '''
+Version $flutterVersionArg not found. Please $openIssueLink to request a new version.
+Use `shorebird flutter versions list` to list available versions.
+''',
+          exitCode: ExitCode.software,
+        );
+      }
+
+      return revision;
+    }
+
+    return shorebirdEnv.flutterRevision;
+  }
+
+  /// Builds the release artifacts for the given platform. Returns the "primary"
+  /// artifact for the platform (e.g. the AAB for Android, the IPA for iOS).
+  Future<FileSystemEntity> buildReleaseArtifacts();
+
+  /// Extracts the release version from the compiled artifact.
+  ///
+  /// We extract the release version from the compiled artifact because we can
+  /// be 100% certain that the artifact will report that same number when making
+  /// patch check requests.
+  Future<String> getReleaseVersion({
+    required FileSystemEntity releaseArtifactRoot,
+  });
+
+  /// Asserts that a release with version [version] can be released using
+  /// flutter version [flutterVersion]. If a release has already been published
+  /// with the given [version] for the platform associated with [releaseType],
+  /// or if a release already exists with [version] but was compiled with a
+  /// different Flutter revision, an error will be thrown.
+  Future<void> validateVersionIsReleasable({
```

Comment:
Nit: when there are methods that are expected to just execute if everything is ok, or to break if there is something inconsistent. I usually like to name it `ensure` or `assert`, like `ensureVersionIsReleaseable` as this makes it more explict (IMO) that an exception will be throws if something is wrong

Reply:
Done. Decided to go with "assert" throughout